File: macos/TokendialCore/Sources/TokendialCore/Alerts/AlertCoordinator.swift
import Foundation

/// The host side of the alert engine: turns readings and sessions into events, runs the reducer under
/// one lock, persists its state, and hands alerts to the sink after filtering the kinds the user switched off.
public final class AlertCoordinator {
    private let lock = NSLock()
    private let sink: AlertSink
    private let now: () -> Date
    private let stateFile: URL?
    private var engine: AlertEngine
    private var wants: (AlertKind) -> Bool
    private var known: [String: String] = [:]
    private var timer: DispatchSourceTimer?

    public init(sink: AlertSink, config: AlertConfig = .default, stateFile: URL? = nil, now: @escaping () -> Date = Date.init, wants: @escaping (AlertKind) -> Bool = { _ in true }) {
        self.sink = sink
        self.now = now
        self.stateFile = stateFile
        self.wants = wants
        engine = Self.load(stateFile, config)
        apply(.restart(at: now()))
    }

    public static var defaultStateFile: URL { Paths.appSupport.appendingPathComponent("alerts.json") }

    public func start(tick: TimeInterval = 15) {
        let timer = DispatchSource.makeTimerSource(queue: .global(qos: .utility))
        timer.schedule(deadline: .now() + tick, repeating: tick)
        timer.setEventHandler { [weak self] in guard let self else { return }; self.apply(.tick(at: self.now())) }
        timer.resume()
        self.timer = timer
    }

    public func stop() { timer?.cancel(); timer = nil }

    /// Thresholds or switches changed: the engine keeps its epochs and only the rules move.
    public func reconfigure(_ config: AlertConfig, wants: @escaping (AlertKind) -> Bool) {
        lock.withLock {
            engine = AlertEngine.load(engine.save(), config: config)
            self.wants = wants
        }
    }

    public func onReadings(_ readings: [ProviderReading]) {
        let at = now()
        for reading in readings where reading.status == .live {
            let limited = reading.block != nil
            for window in reading.windows {
                guard let used = window.usedFraction else { continue }
                let isHeadline = reading.headline?.id == window.id
                apply(.usage(at: at, provider: reading.providerId, window: window.id, usedPct: (used * 10000).rounded() / 100, resetsAt: window.resetsAt, limited: limited && isHeadline))
            }
        }
    }

    public func onActivities(_ activities: [String: Activity]) {
        let at = now()
        var seen: [String: String] = [:]
        for (provider, activity) in activities {
            for session in activity.sessions {
                seen[session.id] = provider
                let state: SessionActivity
                switch session.state { case .waiting: state = .waiting; case .working: state = .working; case .idle: state = .done }
                apply(.session(at: at, provider: provider, sessionId: session.id, state: state))
            }
        }
        let gone = lock.withLock { known.filter { seen[$0.key] == nil } }
        for (id, provider) in gone { apply(.session(at: at, provider: provider, sessionId: id, state: .done)) }
        lock.withLock { known = seen }
    }

    public func onHover(_ on: Bool) { apply(.hover(at: now(), on: on)) }
    public func tick() { apply(.tick(at: now())) }

    private func apply(_ event: AlertEvent) {
        let alerts: [Alert] = lock.withLock {
            let out = engine.reduce(event).filter { wants($0.kind) }
            switch event {
            case .usage, .restart: persist()
            default: if !out.isEmpty { persist() }
            }
            return out
        }
        for alert in alerts {
            Log.alerts.info("\(alert.kind.rawValue) \(alert.provider)\(alert.window.map { "/" + $0 } ?? "")\(alert.pct.map { " \($0)%" } ?? "")")
            sink.deliver(alert)
        }
    }

    private func persist() {
        guard let stateFile else { return }
        do {
            try FileManager.default.createDirectory(at: stateFile.deletingLastPathComponent(), withIntermediateDirectories: true)
            try engine.save().write(to: stateFile, options: .atomic)
        } catch { Log.alerts.error("persist: \(error.localizedDescription)") }
    }

    private static func load(_ file: URL?, _ config: AlertConfig) -> AlertEngine {
        if let file, let data = try? Data(contentsOf: file) { return AlertEngine.load(data, config: config) }
        return AlertEngine(config: config)
    }
}

File: windows/Tokendial.Core/Alerts/AlertCoordinator.cs
using Tokendial.Core.Diagnostics;
using Tokendial.Core.Model;
using Tokendial.Core.Sessions;

namespace Tokendial.Core.Alerts;

/// <summary>
/// The host side of the alert engine: turns readings and sessions into events,
/// runs the reducer under one lock, persists its state, and hands alerts to the
/// sink after filtering the kinds the user switched off. Time comes from a clock
/// the tests can drive.
/// </summary>
public sealed class AlertCoordinator : IDisposable
{
    private readonly object gate = new();
    private readonly IAlertSink sink;
    private readonly Func<DateTimeOffset> now;
    private readonly string? stateFile;
    private AlertEngine engine;
    private Func<AlertKind, bool> wants;
    private Timer? ticker;

    public AlertCoordinator(IAlertSink sink, AlertConfig? config = null, string? stateFile = null, Func<DateTimeOffset>? now = null, Func<AlertKind, bool>? wants = null)
    {
        this.sink = sink;
        this.now = now ?? (() => DateTimeOffset.UtcNow);
        this.stateFile = stateFile;
        this.wants = wants ?? (_ => true);
        engine = LoadEngine(stateFile, config ?? AlertConfig.Default);
        Log.Alerts.Info($"state: {engine.EpochCount} epoch(s) from {(stateFile is null ? "memory" : Path.GetFileName(stateFile))}");
        Apply(new AlertEvent.Restart(this.now()));
    }

    public static string DefaultStateFile => Paths.In("alerts.json");

    public IReadOnlyList<Alert> Delivered { get { lock (gate) return delivered.ToArray(); } }
    private readonly List<Alert> delivered = new();

    public void Start(TimeSpan? tick = null)
    {
        var period = tick ?? TimeSpan.FromSeconds(15);
        ticker = new Timer(_ => Apply(new AlertEvent.Tick(now())), null, period, period);
    }

    /// <summary>Thresholds or switches changed: the engine keeps its epochs and only the rules move.</summary>
    public void Reconfigure(AlertConfig config, Func<AlertKind, bool> wants)
    {
        lock (gate)
        {
            engine = AlertEngine.Load(engine.Save(), config);
            this.wants = wants;
        }
    }

    public void OnReadings(IEnumerable<ProviderReading> readings)
    {
        var at = now();
        foreach (var reading in readings)
        {
            if (reading.Status is not ReadingStatus.Live) continue;
            if (reading.Headline is not UsageWindow window || window.UsedFraction is not double used) continue;
            Apply(new AlertEvent.Usage(at, reading.ProviderId, window.Id, Math.Round(used * 100, 2), window.ResetsAt, reading.Block is not null));
        }
    }

    public void OnActivities(IReadOnlyDictionary<string, Activity> activities)
    {
        var at = now();
        var seen = new HashSet<string>();
        foreach (var (provider, activity) in activities)
        {
            foreach (var session in activity.Sessions)
            {
                seen.Add(session.Id);
                Apply(new AlertEvent.Session(at, provider, session.Id, session.State switch
                {
                    SessionState.Waiting => SessionActivity.Waiting,
                    SessionState.Working => SessionActivity.Working,
                    _ => SessionActivity.Done
                }));
            }
        }
        IReadOnlyList<(string Provider, string SessionId)> waiting;
        lock (gate) waiting = engine.Waiting;
        foreach (var (provider, id) in waiting.Where(w => !seen.Contains(w.SessionId)))
            Apply(new AlertEvent.Session(at, provider, id, SessionActivity.Done));
    }

    public void OnHover(bool on) => Apply(new AlertEvent.Hover(now(), on));

    public void Tick() => Apply(new AlertEvent.Tick(now()));

    private void Apply(AlertEvent e)
    {
        List<Alert> alerts;
        lock (gate)
        {
            alerts = engine.Reduce(e).Where(a => wants(a.Kind)).ToList();
            if (alerts.Count > 0 || e is AlertEvent.Usage or AlertEvent.Restart) Persist();
            delivered.AddRange(alerts);
            if (delivered.Count > 200) delivered.RemoveRange(0, delivered.Count - 200);
        }
        foreach (var alert in alerts)
        {
            Log.Alerts.Info($"{alert.Kind} {alert.Provider}{(alert.Window is null ? "" : "/" + alert.Window)}{(alert.Pct is null ? "" : $" {alert.Pct}%")}");
            try { sink.Deliver(alert); }
            catch (Exception error) { Log.Alerts.Error($"sink: {error.Message}"); }
        }
    }

    private void Persist()
    {
        if (stateFile is null) return;
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(stateFile)!);
            var temp = stateFile + ".tmp";
            File.WriteAllText(temp, engine.Save());
            File.Move(temp, stateFile, overwrite: true);
        }
        catch (Exception error) { Log.Alerts.Error($"persist: {error.Message}"); }
    }

    private static AlertEngine LoadEngine(string? file, AlertConfig config)
    {
        try
        {
            if (file is not null && File.Exists(file)) return AlertEngine.Load(File.ReadAllText(file), config);
        }
        catch (Exception error) { Log.Alerts.Error($"load: {error.Message}"); }
        return new AlertEngine(config);
    }

    public void Dispose() => ticker?.Dispose();
}
